ai21_tokenizer/base_jurassic_tokenizer.py

```diff
@@ -0,0 +1,255 @@
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Dict, Any, BinaryIO, Tuple, Union
+
+import sentencepiece as spm
+
+from abc import ABC
+from ai21_tokenizer.utils import is_number, PathLike, load_json
+
+_MODEL_EXTENSION = ".model"
+_MODEL_CONFIG_FILENAME = "config.json"
+
+
+@dataclass
+class SpaceSymbol:
+    tok_id: int
+    count: int
+
+
+class BaseJurassicTokenizer(ABC):
+    _sp: spm.SentencePieceProcessor = None
+    _id_to_token_map: Dict[int, str] = None
+    _token_to_id_map: Dict[str, int] = None
+    _no_show_tokens: set[str] = None
+    _newline_id: int = 0
+    _space_tokens: List[SpaceSymbol] = None
+
+    def __init__(
+        self,
+        model_path: Optional[PathLike] = None,
+        config: Optional[Dict[str, Any]] = None,
+        model_file_handle: Optional[BinaryIO] = None,
+        model_proto: Optional[bytes] = None,
+    ):
+        self._model_path = model_path
+        self._model_file_handle = model_file_handle
+        self._validate_init(model_path=model_path, model_file_handle=model_file_handle)
+
+        config = self._get_config(model_path=model_path, config=config)
+
+        self.pad_id = config.get("pad_id")
+        self.unk_id = config.get("unk_id")
+        self.eop_id = config.get("eop_id")
+        self.bos_id = config.get("bos_id")
+        self.eos_id = config.get("eos_id")
+
+        self._newline_piece = config.get("newline_piece")
+        self._mask_pieces = config.get("mask_pieces", [])
+
+        self._manual_add_dummy_prefix = not (config.get("add_dummy_prefix", True))
+
+        self._sample_split = re.compile(r"▁*[^▁]+|▁")
+        self._space_split = re.compile("( {2,})")  # Split by 2 or more consecutive spaces
+
+        self._number_mode = config.get("number_mode")
+        self._space_mode = config.get("space_mode")
+
+        if model_proto:
```

Comment:
Is this needed? This line is happening in every one of its child class

Reply:
Removed :)